src/BCFKeyValueData.cc

```diff
@@ -568,6 +577,172 @@ Status BCFKeyValueData::dataset_range(const string& dataset,
+            // now just returning empty results for each dataset.
+            return Status::OK();
+        }
+
+        // advance the KeyValue iterator to the desired dataset
+        string key, key_dataset, value;
+        do {
+            s = it_->next(key, value);
+            if (s.bad()) {
+                if (s == StatusCode::NOT_FOUND) {
+                    // NOT_FOUND here indicates we've reached the end of the
+                    // whole database!
+                    it_.reset();
+                    return Status::OK();
+                }
+                return s;
+            }
+
+            // parse the key
+            string key_bucket;
+            s = body_.rangeHelper->parse_key(key, key_bucket, key_dataset);
+            if (s.bad()) {
+                return s != StatusCode::NOT_FOUND ? move(s) : Status::Failure("BCFBucketIterator::next(): failed to parse key", key);
+            }
+
+            if (key_bucket != bucket_) {
+                // we've advanced past the end of the bucket
+                assert(key_bucket > bucket_);
+                it_.reset();
+                return Status::OK();
+            }
+        } while (key_dataset < dataset); // repeat until we reach the desired dataset
+        assert(key_dataset == dataset);
+
+        // extract the records overlapping range_
+        s = scan_bucket(dataset, value, hdr.get(), range_, stats_, records);
+        if (s.ok()) {
+            stats_.nBCFRecordsInRange += records.size();
+        }
+        return s;
+    }
+};
+
+Status BCFKeyValueData::sampleset_range(const MetadataCache& metadata, const string& sampleset,
+                                        const range& pos,
+                                        shared_ptr<const set<string>>& samples,
+                                        shared_ptr<const set<string>>& datasets,
+                                        vector<unique_ptr<RangeBCFIterator>>& iterators) {
+    Status s;
+
+    // resolve samples and datasets
+    S(metadata.sampleset_datasets(sampleset, samples, datasets));
+
+    // TODO: dispatch to sampleset_range_base if the sample set is "too
+    // small". need a way to know N of the whole database
+
+    // get a KeyValue::Reader so that all iterators read from the same
+    // snapshot (this isn't strictly necessary since datasets are immutable,
+    // but seems nice to have)
+    unique_ptr<KeyValue::Reader> ureader;
+    S(body_->db->current(ureader));
+    shared_ptr<KeyValue::Reader> reader(move(ureader));
+
+    // create one iterator per bucket extent
+    shared_ptr<BucketExtent> bkExt = body_->rangeHelper->scan("", pos);
+    iterators.clear();
+    for (range r = bkExt->begin(); r <= bkExt->end(); r = bkExt->next()) {
+        // form a key less than all keys in this bucket
+        string key0 = body_->rangeHelper->gen_key("", r);
+        string bucket, blank;
+        S(body_->rangeHelper->parse_key(key0, bucket, blank));
+        assert(blank.empty());
+
```

Comment:
Suggest creating a special function in the rangeHelper that will extract the bucket from a range. It could simply be the above four lines packed into one function.